File: src/dataset/python/linked-list_exercism-ThomasZumsteg/gemini/GeminiFlash_linked_list_v2.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None
        self.prev = None


class LinkedList:
    def __init__(self):
        self.head = None
        self._len = 0

    def __len__(self):
        return self._len

    def __iter__(self):
        if self.head is None:
            return
        node = self.head
        yield node.value
        node = node.next
        while node is not self.head and node is not None:
            yield node.value
            node = node.next

    def push(self, value):
        new_node = Node(value)
        if self.head is None:
            self.head = new_node
            new_node.next = new_node
            new_node.prev = new_node
        else:
            tail = self.head.prev
            new_node.next = self.head
            new_node.prev = tail
            self.head.prev = new_node
            tail.next = new_node
            self.head = new_node
        self._len += 1

    def pop(self):
        if self.head is None:
            raise IndexError("pop from an empty LinkedList")

        value = self.head.value
        if self.head.next == self.head:
            self.head = None
        else:
            tail = self.head.prev
            self.head = self.head.next
            self.head.prev = tail
            tail.next = self.head
        self._len -= 1
        return value

    def shift(self):
        if self.head is None:
            raise IndexError("shift from an empty LinkedList")

        tail = self.head.prev
        value = tail.value

        if self.head.next == self.head:
            self.head = None
        else:
            self.head.prev = tail.prev
            tail.prev.next = self.head
        self._len -= 1
        return value

    def unshift(self, value):
        new_node = Node(value)
        if self.head is None:
            self.head = new_node
            new_node.next = new_node
            new_node.prev = new_node
        else:
            tail = self.head.prev
            new_node.next = self.head
            new_node.prev = tail
            self.head.prev = new_node
            tail.next = new_node
            self.head = new_node

        self._len += 1

File: src/dataset/javascript/linked-list_exercism-ThomasZumsteg/gemini/GeminiFlash_linked-list_v2.js
var LinkedList = function() {
	this.head = null;
	this.tail = null;
	this.length = 0;
};

LinkedList.prototype.push = function(val) {
	const newElement = { val: val, next: null, prev: this.tail };

	if (this.tail) {
		this.tail.next = newElement;
	} else {
		this.head = newElement;
	}

	this.tail = newElement;
	this.length++;
};

LinkedList.prototype.pop = function() {
	if (!this.head) {
		return undefined;
	}

	const element = this.tail;
	if (this.length === 1) {
		this.head = null;
		this.tail = null;
	} else {
		this.tail = element.prev;
		this.tail.next = null;
	}

	this.length--;
	return element.val;
};

LinkedList.prototype.unshift = function(val) {
	const newElement = { val: val, next: this.head, prev: null };

	if (this.head) {
		this.head.prev = newElement;
	} else {
		this.tail = newElement;
	}

	this.head = newElement;
	this.length++;
};

LinkedList.prototype.shift = function() {
	if (!this.head) {
		return undefined;
	}

	const element = this.head;
	if (this.length === 1) {
		this.head = null;
		this.tail = null;
	} else {
		this.head = element.next;
		this.head.prev = null;
	}

	this.length--;
	return element.val;
};

LinkedList.prototype.count = function() {
	return this.length;
};

LinkedList.prototype.delete = function(val) {
	let current = this.head;

	while (current) {
		if (current.val === val) {
			if (current.prev) {
				current.prev.next = current.next;
			} else {
				this.head = current.next;
			}

			if (current.next) {
				current.next.prev = current.prev;
			} else {
				this.tail = current.prev;
			}

			this.length--;
			return;
		}
		current = current.next;
	}
};

module.exports = LinkedList;